examples/de-emphasize.rs

```diff
@@ -0,0 +1,88 @@
+extern crate mdbook;
+extern crate pulldown_cmark;
+extern crate pulldown_cmark_to_cmark;
+
+// This program removes all forms of emphasis from the markdown of the book.
+
+use mdbook::errors::Error;
+use mdbook::MDBook;
+use mdbook::book::{Book, BookItem};
+use mdbook::preprocess::{Preprocessor, PreprocessorContext};
+use pulldown_cmark::{Event, Parser, Tag};
+use pulldown_cmark_to_cmark::fmt::cmark;
+
+use std::ffi::OsString;
+use std::env::{args, args_os};
+use std::process;
+
+struct Deemphasize;
+
+impl Preprocessor for Deemphasize {
+    fn name(&self) -> &str {
+        "md-links-to-html-links"
+    }
+
+    fn run(&self, _ctx: &PreprocessorContext, book: &mut Book) -> ::std::result::Result<(), Error> {
+        eprintln!("Running '{}' preprocessor", self.name());
+        let mut res: Option<_> = None;
+        let mut num_removed_items = 0;
+        book.for_each_mut(|item: &mut BookItem| {
+            if let &Some(Err(_)) = &res {
+                return;
+            }
+            if let BookItem::Chapter(ref mut chapter) = *item {
+                eprintln!("{}: processing chapter '{}'", self.name(), chapter.name);
+                let mut buf = String::with_capacity(chapter.content.len());
+                res = Some({
+                    let events = Parser::new(&chapter.content).filter(|e| {
+                        let should_keep = match *e {
+                            Event::Start(Tag::Emphasis)
+                            | Event::Start(Tag::Strong)
+                            | Event::End(Tag::Emphasis)
+                            | Event::End(Tag::Strong) => false,
+                            _ => true,
+                        };
+                        if !should_keep {
+                            num_removed_items += 1;
+                        }
+                        should_keep
+                    });
+                    cmark(events, &mut buf, None)
+                });
+                if let &Some(Ok(_)) = &res {
+                    chapter.content = buf;
+                }
+            }
+        });
+        eprintln!(
+            "{}: removed {} events from markdown stream.",
+            self.name(),
+            num_removed_items
+        );
+        if let Some(Err(err)) = res {
+            Err(Error::from(format!(
+                "Markdown serialization failed: {}",
+                err
+            )))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+fn do_it(book: OsString) -> ::std::result::Result<(), Error> {
```

Comment:
Just out of curiosity, what was the reasoning for explicitly take an `OsString` here? It's not every day you see `OsString` being used outside of an FFI/OS context.

We also load books from the file system, so I usually take a `&Path` or even `P: Into<PathBuf>` or `P: AsRef<Path>` like in [`File::open()`][1].

[1]: https://doc.rust-lang.org/nightly/std/fs/struct.File.html#method.open

Reply:
Arguments are natively stored in `OsStrings` which have arbitrary encoding. `Paths` also assume no encoding, as they are interpreted by the OS.
Thus I generally don't convert program arguments to `String` if they should end up in a `Path` as this would unnecessarily attempt to decode them as utf-8, which might fail.